docs/question/server.md

```diff
@@ -1,17 +1,104 @@
-| `file()`     | `object` (string, bytes-like, file-like) |                                                                                                          | `correct_answers`, `filename`, `options`, `params`, `variant_seed`                                                                                                                                         | Generate a file object dynamically in lieu of a physical file. Trigger via `type="dynamic"` in the question element (e.g., `pl-figure`, `pl-file-download`). Access the requested filename via `data['filename']`. If `file()` returns nothing, an empty string will be used. |
+## Guide
 
-A complete `question.html` and `server.py` example looks like:
+To explain how this works, we will use a simple example of a question that asks the student to double a number. The question will look like this:
+
+<!-- prettier-ignore -->
+```html title="question.html"
+<pl-question-panel>
+  If $x = {{params.x}}$, what is $y$ if $y$ is double $x$?
+</pl-question-panel>
+<pl-number-input answers-name="y" label="$y =$"></pl-number-input>
+<pl-submission-panel>
+  {{feedback.y}}
+</pl-submission-panel>
+```
+
+More details about the `{{params.x}}` and `{{feedback.y}}` Mustache syntax can be found in the [`question.html` documentation](./html.md).
+
+### Step 1: `generate`
+
+First, the `generate` function is called to generate random parameters for the variant, and the correct answers. It should set `data["params"]` with the parameters for the question, and `data["correct_answers"]` with the correct answers. The parameters can then be used in the `question.html` file by using `{{params.NAME}}`.
+
+```python title="server.py"
+import random
+
+def generate(data):
+    # Generate random parameters for the question and store them in the data["params"] dict:
+    data["params"]["x"] = random.randint(5, 10)
+
+    # Also compute the correct answer (if there is one) and store in the data["correct_answers"] dict:
+    data["correct_answers"]["y"] = 2 * data["params"]["x"]
+```
+
+!!! info
+
+    In general, each function of the question generation process runs *after* all elements in the question. For example, the `generate()` function in `server.py` runs after all elements have run their `generate()` functions. This is important to remember when using the `data` dictionary, as it will contain the results *after* the elements have finished processing.
```

Comment:
I think the placement of this info box is a bit misleading, since none of the provided elements use a `generate` function. I'd suggest moving this either to the next step or before all steps. In the latter case, `parse()` and `grade()` might be better examples since they are often combined (element+question) in cases where custom grading is used.

Reply:
I moved the info box to `parse` in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9.